durable-storage/src/persistence_layer.rs

```diff
@@ -0,0 +1,141 @@
+//! - clone
+//! - commit (returning the commit hash)
+//! - checkout a specific commit
+
+use std::mem::ManuallyDrop;
+use std::path::PathBuf;
+
+use crate::repo::DirectoryManager;
+
+/// Type alias for a 32-byte hash used for identifying key-value blobs & commits.
+type Hash = [u8; 32];
+
+/// Errors encountered when interacting with the persistence layer.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("RocksDB error: {0}")]
+    RocksDBError(#[from] rocksdb::Error),
+}
+
+/// Create a [`rocksdb::Options`] instance for creating a new database.
+fn options_expect_newly_created() -> rocksdb::Options {
+    // To avoid accidentaly overwriting an existing database, we set `error_if_exists`.
+    let mut options = rocksdb::Options::default();
+    options.create_if_missing(true);
+    options.set_error_if_exists(true);
+    options
+}
+
+/// Persistence layer for durable solution used by the RISC-V PVM.
+///
+/// Invariants:
+/// - The path in `temp_initial_db_path` is unique for each instance of [`PersistenceLayer`] and is
+///   assumed to not be modified / known outside of this instance.
+pub struct PersistenceLayer {
+    /// The underlying handle to the RocksDB instance.
+    ///
+    /// [`ManuallyDrop`] is used to ensure safety when dropping [`PersistenceLayer`]. Calling
+    /// [`rocksdb::DB::destroy`] requires all connections to that path to be closed, which is called
+    /// in [`rocksdb::DB`]'s drop method.
+    db_instance: ManuallyDrop<rocksdb::DB>,
+
+    /// If the current [`PersistenceLayer`] was created with `new` or `clone`, this holds the path
+    /// to the temporary directory created within the repository.   
+    temp_initial_db_path: Option<PathBuf>,
+}
+
+impl PersistenceLayer {
+    /// Creates a new `PersistenceLayer` instance within the given `repo`.
+    pub fn new(repo: &DirectoryManager) -> Result<Self, Error> {
+        let new_db_path = repo.new_checkpoint_dir();
+
+        let db = rocksdb::DB::open(&options_expect_newly_created(), &new_db_path)?;
+
+        Ok(Self {
+            temp_initial_db_path: Some(new_db_path),
+            db_instance: ManuallyDrop::new(db),
+        })
+    }
+}
+
+impl Drop for PersistenceLayer {
+    /// Databases created from a new or clone operation will have `temp_initial_db_path` set. These
+    /// databases do not have to be saved on disk as they have not been committed to storage.
+    fn drop(&mut self) {
+        // Safety: This manual drop is called in this object's drop method.
+        unsafe {
+            ManuallyDrop::drop(&mut self.db_instance);
+        }
+
+        // SAFETY: Although marked as safe, destroy on a path requires all rocksdb connections to
+        // this path to be closed. This is why we need to manual drop the db_instance first & the
+        // invariants of `PersistenceLayer` to be upheld.
+        if let Some(path) = &self.temp_initial_db_path {
+            let _ = rocksdb::DB::destroy(&rocksdb::Options::default(), path);
```

Comment:
should this error ever occur? it might be worth logging if it does and/or panicking (if not already panicking?)

Reply:
> For logging the error I'd need to depend on the PVM library
Why?